### PR TITLE
Update device authentication

### PR DESCRIPTION
**Pull request by gferraro**
Will supply deviceName and groupName to authenticate_device API

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -113,7 +114,7 @@ func NewAPIFromConfig(configFile string) (*CacophonyAPI, error) {
 			return nil, err
 		}
 	}
-	api, err := NewAPI(conf.ServerURL, conf.Group, conf.DeviceName, password)
+	api, err := NewAPI(conf.ServerURL, conf.Group, conf.DeviceName, 0, password)
```
Comment (mjs) [inline api.go]: Should this be passing the device id from the config file if it's available? Ultimately we want to just authenticate with the device id right? Then only registration needs to have the group and device name.

File: api.go [hunk 2]
```diff
@@ -167,11 +175,14 @@ func (api *CacophonyAPI) authenticate() error {
 	if api.device.password == "" {
 		return errors.New("no password set")
 	}
-	payload, err := json.Marshal(map[string]string{
+
+	payload, err := json.Marshal(map[string]interface{}{
+		"deviceID":   api.device.id,
```
Comment (mjs) [inline api.go]: how about you only add `deviceID` if it isn't `0`

File: api.go [hunk 3]
```diff
@@ -167,11 +175,14 @@ func (api *CacophonyAPI) authenticate() error {
...
 		"devicename": api.device.name,
 		"groupname":  api.device.group,
 		"password":   api.device.password,
```
Comment (mjs) [inline api.go]: Is the API server `authenticate_device` endpoint already able to support logins with just the device id? If so, perhaps don't send the group name and password if there's a device id available?

File: api_test.go [hunk 4]
```diff
@@ -181,6 +183,7 @@ func TestAPIRegistration(t *testing.T) {
 	assert.NoError(t, err)
 	assert.True(t, api.JustRegistered())
 	assert.NotEqual(t, "", api.device.password)
+	// assert.NotEqual(t, 0, api.device.id)
```
Comment (mjs) [inline api_test.go]: is this supposed to be commented out?

File: config.go [hunk 5]
```diff
@@ -53,6 +54,17 @@ func (conf *Config) Validate() error {
 	return nil
 }
 
+//SaveConfigFile writes config to supplied filename
+func (conf *Config) SaveConfigFile(filename string) error {
```
Comment (mjs) [inline config.go]: Give that it's already a method on `Config`, this could just be called `SaveFile` or `WriteFile`.

File: config.go [hunk 6]
```diff
@@ -53,6 +54,17 @@ func (conf *Config) Validate() error {
...
+	buf, err := yaml.Marshal(&conf)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(filename, buf, 0600)
+	return err
```
Comment (mjs) [inline config.go]: Just `return ioutil.WriteFile(filename, buf, 0600)` :)

File: api_test.go (hunk 4 above)
Comment (gferraro) [reply api_test.go]: yea as wont be a value until cacophony-api/master is updated

File: api.go (hunk 1 above)
Comment (gferraro) [reply api.go]: ofcourse :) silly me

File: api.go (hunk 3 above)
Comment (gferraro) [reply api.go]: password is always needed

**Comment by gferraro** [comment]
Updated as per comments

File: api_test.go [hunk 7]
```diff
@@ -78,11 +78,12 @@ func getTokenResponse() *tokenResponse {
...
 		Messages: []string{},
 		Token:    "tok-" + randString(20),
+		ID:       1,
 	}
 }
 
-func getJSONRequestMap(r *http.Request) map[string]string {
-	var requestJson = map[string]string{}
+func getJSONRequestMap(r *http.Request) map[string]interface{} {
+	var requestJson = map[string]interface{}{}
```
Comment (mjs) [inline api_test.go]: You don't need to create an instance here as Decode will do that. Just `var requestJson map[string]interface{}` is fine.

**Pull request merged**